Scheduler modes (#477)
kubos/kubos: docs/ecosystem/services/scheduler.rst

Scheduler
=========

The KubOS system includes a scheduler service to facilitate recurring
and one time tasks with specific timing requirements.

Behavior
--------

The behavior of the scheduler is defined through two levels of organization: `schedules`
and `modes`. A `schedule` is a file containing a list of tasks and the relevant information
for scheduling and executing the task. A `mode` is a directory which contains many `schedules`.

Upon boot, or service start, the scheduler will read the active mode directory and
load all schedules from that mode into memory. The default active mode directory
will be found at ``/home/system/etc/schedules/active``. 

Schedules consist of three types of tasks: ``init``, ``one-time``, and
``recurring``. Any ``init`` tasks will be performed immediately. All other 
``one-time`` or ``recurring`` tasks will be scheduled to run at their designated times.

By default the scheduler comes with a single mode: ``safe``. This mode is reserved as a
fallback and default operating mode. Additional modes can be created and made active
through the GraphQL interface. All modes are represented as different directories in

> the schedules directory. Only one mode can be active at any given time.

Reviewer: > in the schedules directory *(`/home/system/etc/schedules`)*.

> The scheduler will have its own log file, ``/var/log/kubos-schedule.log``, which

Reviewer: > The scheduler *has* its own log file,

> which logs all schedule related actions

Reviewer: More *will* below as well that should be declarative.

Scheduler
=========

The KubOS system includes a scheduler service to facilitate recurring
and one time tasks with specific timing requirements.

Behavior
--------

The behavior of the scheduler is defined through two levels of organization: *schedules*
and *modes*. A *schedule* is a file containing a list of tasks and the relevant information
for scheduling and executing the task. A *mode* is a directory which contains many *schedules*.

Upon boot, or service start, the scheduler reads the active mode directory and
load all schedules from that mode into memory. The default active mode directory

> is found at ``/home/system/etc/schedules/active``. 

Reviewer 2: Nit: Might be worth mentioning that this is a symlink to the active schedule directory

Schedules consist of three types of tasks: ``init``, ``one-time``, and

> ``recurring``. Any ``init`` tasks will be performed immediately. All other 

Reviewer 2: nit: Could use an expansion on what "immediately" is (ie when the schedule becomes active, or when the scheduler service is started). I know it's explained later on, but it would also be nice here.

``one-time`` or ``recurring`` tasks will be scheduled to run at their designated times.

By default the scheduler comes with a single mode: ``safe``. This mode is reserved as a
fallback and default operating mode. Additional modes can be created and made active
through the GraphQL interface. All modes are represented as different directories in
the schedules directory (``/home/system/etc/schedules``).
Only one mode can be active at any given time.

The scheduler has its own log file, ``/var/log/kubos-schedule.log``, which
logs all schedule related actions the scheduler takes.

Schedule Specification
----------------------

Schedules are specified in files in the `json` format. Each schedule file contains

> all of the necessary information for that specific schedule. Different schedule files can

Reviewer 3: ```suggestion
all of the necessary information for each task within the schedule. Different schedule files can
```

Scheduler
=========

The KubOS system includes a scheduler service to facilitate recurring
and one time tasks with specific timing requirements.

Behavior
--------

The behavior of the scheduler is defined through three levels of organization: *config*

> , *mode*, and *schedule*. A *config* is a file containing a list of tasks and the

Author: Yea, I wasn't feeling totally thrilled with config files, that was just the first alternative to schedule file that came to my mind. I'd be ok with task files. Any thoughts @jacoffey3?

> , *mode*, and *schedule*. A *config* is a file containing a list of tasks and the

Reviewer 3: I'd be okay with that. This whole terminology bit is quite difficult to get right. 

Try testing it out on Marshall or Tyler?

> , *mode*, and *schedule*. A *config* is a file containing a list of tasks and the

Reviewer 3: Maybe just gloss over the files part in the organization, but mention it as an implementation detail? 

```
The behavior of the scheduler is defined through three levels of organization: *mode*, *schedule*, and *task*. The *mode* determines what schedule directory is being executed. The *schedule* is then assembled through combining all the *task files* in the mode directory, each of which can contain one or many *tasks*. 
```

> , *mode*, and *schedule*. A *config* is a file containing a list of tasks and the

Reviewer 3: Now that I typed that out I like it a lot better

> , *mode*, and *schedule*. A *config* is a file containing a list of tasks and the

Author: We talked to Tyler yesterday and he suggested we refer to the files as _task lists_. The _schedule_ of the _mode_ is then made up of the aggregate _tasks_ of all the _task lists_ in the _mode_.

> , *mode*, and *schedule*. A *config* is a file containing a list of tasks and the

Reviewer 3: That's good! So: 

"The behavior of the scheduler is defined through three levels of organization: _mode_, _schedule_, and _task_. The _mode_ determines what schedule directory is being executed. The _schedule_ is then assembled through combining all the _task lists_ in the schedule directory, each of which can contain one or many _tasks_. "

> , *mode*, and *schedule*. A *config* is a file containing a list of tasks and the

Author: New verbage has been addressed in #479

relevant information for scheduling and executing the task. A *mode* is a directory which
contains many *config* files. The contents of all the *configs* found in a *mode*
make up the *schedule* of that mode.

Upon boot, or service start, the scheduler reads all config files in the active 
mode directory and schedules all tasks from that mode. The default active mode directory
is found at ``/home/system/etc/schedules/active``, which is a symlink
to the actual active mode's directory.

Schedules consist of three types of tasks: ``init``, ``one-time``, and
``recurring``. Tasks are loaded into the scheduler when the scheduler service starts,
when a mode becomes active, or when a new schedule file is imported into the active mode.
Any ``init`` tasks will be scheduled for execution when their corresponding
delay has passed. All other ``one-time`` or ``recurring`` tasks will be scheduled to
run at their designated times.

> By default the scheduler comes with a single mode: ``safe``. This mode is reserved as a

Reviewer 2: nit: capitalization - `SAFE`. If you try to register a schedule to `safe` mode it fails because `{schedule-dir}/safe` doesn't exist.

Tangent: It might be useful to allow `SAFE` or `safe` to be specified

Reviewer 3: I think modes should be case insensitive I think.

Author: We can make them case insensitive by converting mode names to either upper or lower case inside of the scheduler.  Do we have a preference? I'm actually leaning towards moving `SAFE` to `safe` and converting all modes to all lowercase, as all caps isn't really a normal linux folder naming convention.

Reviewer 3: Yeah I like that idea. All lower case lines up with the rest of our system generally

.. code-block:: json

    {
        "name": "Descriptive task-name",
        "delay": "Required start delay in Xh Ym Zs format",
        "period": "Required period of execution in Xh Ym Zs format",
        "app": {
            "name": "Required registered name of app to run",
            "args": ["Optional", "command", "line", "app", "args"],
            "config": "Optional path to app config"
        }
    }

An example schedule config:

.. code-block:: json

    {
        "init": [
            {
                "name": "start-camera",
                "delay": "10m",
                "app": {
                    "name": "activate-camera"
                }
            }
        ],
        "one-time": [
            {
                "name": "deploy-solar",
                "time": "2019-08-11 15:20:10",
                "app": {
                    "name": "deploy-solar-panels"
                }
            }
        ],
        "recurring": [
            {

>                 "name": "clean-logs-every-12hrs":

Reviewer 2: nit: should be a comma at the end of this line, not a colon

specified in the ``config.toml`` file under ``[scheduler-service]``.

- ``schedules-dir`` - (Default: ``/home/system/etc/schedules/``) The path to the

> directory where modes and schedules will be stored. This directory will be created if it does

Reviewer: > modes and *their* schedules will be stored.

Queries
~~~~~~~

> The scheduler exposes a two queries, ``activeMode`` and ``availableModes``.

Reviewer: > The scheduler exposes ~~a~~ two queries

The ``activeMode`` query  exposes information about the currently active

> schedule mode. It has the following schema::

Reviewer: > ~~schedule~~ mode

    {
        activeMode: {
            path: String,
            name: String,
            timeCreated: String,

>             schedules: [Schedule],

Reviewer: When a new schedule is added, I assume the service will reload the directory such that this gets updated?

Author: That is a good question. Should this time represent the last time a mode was updated/reloaded or when it was initially created?

Reviewer 3: Maybe rename it to `lastRevised` and make it the last time a schedule or tasks were added. `timeCreated` isn't as useful

Reviewer: When a schedule is imported, are we reloading the data such that `schedules` will now reflect the new ones?

Reviewer: ^

Author: Yes this is a good question. I have not implemented the reload on import behavior yet. There are two obvious options when a schedule is imported into an active mode:
- Reload and re-schedule all tasks across all files in the mode
- Load and schedule any tasks in the newly imported file
Thoughts @jacoffey3?

Reviewer 3: This one: 

> Only load and schedule any tasks in the newly imported file 

As a slight tangent, would it be possible to just make this mutation either a file path or JSON you pass to it directly? Like to schedule a single item? Then you can just make the file.

Author: The `importSchedule` mutation is already a file path. We could create a new `importTask` mutation which allows you to pass in the task/app parameters and makes a file from it.

Reviewer 3: I would love that. It'd be excellent for putting in 1 off tasks.

            active: Boolean
        }
    }

The ``availableModes`` query  exposes information about the currently available
schedule modes. It has the following schema::

    {
        availableModes(name: String): [
            {
               path: String,
               name: String,
               timeCreated: String,
               schedules: [Schedule],
               active: Boolean
            }
        ]
    }

The ``schedule`` object exposes information about individual schedule files. It
has the following schema::

    {
        schedule:
            {
               contents: String,
               path: String,
               name: String,
               timeImported: String
            }
        ]
    }

Mutations
~~~~~~~~~

The scheduler has the following mutations: ``createMode``, ``removeMode``,
``activateMode``, ``importSchedule``, and ``removeSchedule``.

The ``createMode`` mutation instructs the scheduler to create a new empty schedule mode.
It has the following schema::

    mutation {
        createMode(name: String!) {
            success: Boolean,
            errors: String
        }
    }

> The ``removeMode`` mutation instructs the scheduler to delete an existing schedule mode.

Reviewer: > *It cannot be applied to the currently active mode.*

It has the following schema::

    mutation {
        removeMode(name: String!) {
            success: Boolean,
            errors: String
        }
    }

> The ``activateMode`` mutation instructs the scheduler to make the specified schedule mode

Reviewer: > ~~schedule~~ mode

active. It has the following schema::

    mutation {
        activateMode(name: String!): {
            success: Boolean,
            errors: String
        }
    }

> The ``safeMode`` mutation instructs the scheduler to make the safe mode

Reviewer: Does this need to exist, as opposed to just `activateMode(name: "safe")`?

Author: @jacoffey3 had requested a standalone interface for activating the safe mode.

            timeCreated: String,
            schedules: [Schedule],
            active: Boolean
        }
    }

The ``availableModes`` query  exposes information about the currently available
modes. It has the following schema::

    {
        availableModes(name: String): [
            {
               path: String,
               name: String,
               timeCreated: String,
               schedules: [Schedule],
               active: Boolean
            }
        ]
    }

The ``schedule`` object exposes information about individual schedule files. It
has the following schema::

    {
        schedule:
            {
               contents: String,
               path: String,
               name: String,
               timeImported: String
            }
        ]
    }

Mutations
~~~~~~~~~

The scheduler has the following mutations: ``createMode``, ``removeMode``,
``activateMode``, ``importSchedule``, and ``removeSchedule``.

The ``createMode`` mutation instructs the scheduler to create a new empty schedule mode.
It has the following schema::

    mutation {
        createMode(name: String!) {
            success: Boolean,
            errors: String
        }
    }

> The ``removeMode`` mutation instructs the scheduler to delete an existing schedule mode.

Reviewer 3: It deletes the mode directory and all schedules within.
